### PR TITLE
snowflake_streaming: compare offset tokens numerically, not lexicographically

### PR DESCRIPTION
**Pull request by Jeffail**
hey folks — found a rough edge in `snowflake_streaming`'s exactly-once dedup filter while poking around a related area, thought it was worth fixing rather than just flagging.

## what's going on

`offset_token` accepts any interpolated expression, and `preprocessForExactlyOnce` compares successive tokens as plain Go strings to decide whether a row is already committed. That's only correct if every token is padded to a fixed width — the docs did call this out, but a bare numeric token such as `${!@kafka_offset}` is the most obvious thing to reach for and doesn't meet that requirement, so it was easy to end up relying on a caveat that's simple to miss on a first read.

Unpadded numeric tokens cross a digit-count boundary (9→10, 99→100, …) on any long-running channel, and at that point:

- a genuinely new row compares as "less than" the latest committed token and gets dropped instead of inserted
- conversely, a redelivered already-committed row compares as "greater than" and gets reinserted as a duplicate

## the fix

Added `compareOffsetTokens`, which compares both sides as base-10 integers when they parse as such, falling back to the previous lexicographic comparison otherwise (so non-numeric custom tokens are unaffected). Plain numeric tokens no longer need padding at all — this removes the caveat rather than just documenting it more clearly.

Also split the comparison logic out into `filterAlreadyCommitted` (takes the latest offset token value directly rather than the channel) so it has proper unit test coverage without needing a live `SnowflakeIngestionChannel`, and updated the field's own documentation to describe the simplified rule.

## testing

Added regression tests covering the digit-count boundary case (committed token `"9"`, incoming `"10"`), plus redelivery/dedup and mixed-batch cases. Ran `task fmt`, `task test:unit` clean. `task lint` currently fails on this checkout with an unrelated toolchain issue (`sync/atomic` export data version mismatch) — confirmed it fails the same way on a clean `main` with no changes, so not something this PR introduces.

No rush on review, happy to adjust the approach if there's a preference for how the comparison should live (e.g. exported vs. package-private).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/impl/snowflake/output_snowflake_streaming.go [hunk 1]
```diff
@@ -1221,7 +1232,28 @@ func preprocessForExactlyOnce(
...
-	return preprocessForExactlyOnce(channel, offsetTokenMapping, filteredBatch)
+	return filterAlreadyCommitted(latest, offsetTokenMapping, filteredBatch)
+}
+
+// compareOffsetTokens orders two offset tokens numerically when both parse as
+// base-10 integers (the default `${! @kafka_offset }` case), so that e.g.
+// "10" correctly sorts after "9" across a digit-count boundary. It falls back
+// to a lexicographic byte comparison when either token isn't a plain
+// integer, preserving today's behaviour for non-numeric custom tokens.
+func compareOffsetTokens(a, b string) int {
```
[the remark this answers is not in the document]
Comment (Jeffail) [reply internal/impl/snowflake/output_snowflake_streaming.go:1247]: good catch, thanks — pushed a follow-up commit updating the `offset_token` field description to describe the numeric-first comparison (and when padding is still needed), and regenerated the docs page with `task docs`.

File: internal/impl/snowflake/output_snowflake_streaming.go [hunk 2]
```diff
@@ -187,13 +188,16 @@ NOTE: There is a limit of 10,000 streams per table - if using more than 10k stre
...
 is dropped. This means it is *very important* to have ordered delivery to the output, any out of order messages to the output will be seen as duplicates and dropped.
 Specifically this means that retried messages could be seen as duplicates if later messages have succeeded in the meantime, so in most circumstances a dead letter queue
 output should be employed for failed messages.
 
-NOTE: It's assumed that messages within a batch are in increasing order by offset token, additionally if you're using a numeric value as an offset token, make sure to pad
-      the value so that it's lexicographically ordered in its string representation, since offset tokens are compared in string form.
+Offset tokens that both parse as base-10 integers (up to 64 bits) are compared numerically, so plain incrementing numeric tokens (such as the default Kafka offset) do not need
+padding. Any other tokens, including numbers outside of the 64-bit integer range, fall back to a lexicographic comparison of their string representation, so if you're using a
+non-numeric or oversized-numeric value as an offset token, make sure to pad it so that it's lexicographically ordered in its string representation.
```
[the remark this answers is not in the document]
Comment (Jeffail) [reply internal/impl/snowflake/output_snowflake_streaming.go]: good catch on this too — the "default" framing was a leftover assumption from the v2 connector work (that one does have a default offset_token expression, this one doesn't). Pushed a follow-up reworking the field description and the compareOffsetTokens comment to drop "default" entirely, call out `${!@kafka_offset}` as just an example of a bare numeric token, and make explicit that both of the field's own examples (`offset-...`, `postgres-...`) are non-numeric due to their prefix and so still need padding. Also scrubbed the same "default" wording out of the PR description.

File: internal/impl/snowflake/output_snowflake_streaming.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2024 Redpanda Data, Inc.
+// Copyright 2026 Redpanda Data, Inc.
```
[the remark this answers is not in the document]
Comment (Jeffail) [reply internal/impl/snowflake/output_snowflake_streaming.go]: good spot, thanks — reverted the copyright year back to 2024 to keep the diff scoped to the behaviour change, matching the rest of the package.

**Review by Leward — APPROVED** [review]
